edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -132,63 +132,153 @@ public async Task SetEndpoint(Endpoint newEndpoint, IList<uint> priorities)
             }
         }
 
-        void UpdatePriorities(IList<uint> priorities)
+        public async Task UpdatePriorities(IList<uint> priorities, Option<Endpoint> newEndpoint)
         {
+            Preconditions.CheckArgument(priorities.Count > 0);
+
+            if (this.closed)
+            {
+                throw new InvalidOperationException($"Endpoint executor for endpoint {this.Endpoint} is closed.");
+            }
+
             // Update priorities by merging the new ones with the existing.
             // We don't ever remove stale priorities, otherwise stored messages
             // pending for a removed priority will never get sent.
-            IList<uint> snapshot = this.priorities.Value;
-            this.priorities.Value = priorities
-                .Union(snapshot)
-                .OrderBy(p => p)
-                .ToList();
+            OrderedDictionary snapshot = this.prioritiesToFsms;
+            foreach (uint priority in priorities)
+            {
+                if (!snapshot.Contains(priority))
+                {
+                    string id = GetMessageQueueId(this.Endpoint.Id, priority);
+
+                    // Create a message queue in the store for every priority we have
+                    await this.messageStore.AddEndpoint(id);
+
+                    // Create a checkpointer and a FSM for every message queue
+                    ICheckpointer checkpointer = await this.checkpointerFactory.CreateAsync(id);
+                    EndpointExecutorFsm fsm = new EndpointExecutorFsm(this.Endpoint, checkpointer, this.config);
+
+                    // Add it to our dictionary
+                    snapshot.Add(priority, fsm);
+                }
+                else
+                {
+                    // Update the existing FSM with the new endpoint
+                    var fsm = (EndpointExecutorFsm)snapshot[priority];
+                    newEndpoint.ForEach(async e => await fsm.RunAsync(Commands.UpdateEndpoint(e)));
+                }
+            }
+
+            this.prioritiesToFsms.Value = snapshot;
+            this.lastUsedFsm = (EndpointExecutorFsm)snapshot[0];
+        }
+
+        static string GetMessageQueueId(string endpointId, uint priority)
+        {
+            if (priority == RouteFactory.DefaultPriority)
+            {
+                // We need to maintain backwards compatibility
+                // for existing sequential stores that don't
+                // have the "_Pri<x>" suffix. We use the default
+                // priority (2,000,000,000) for this, which means
+                // the store ID is just the endpoint ID.
+                return endpointId;
+            }
+            else
+            {
+                // The actual ID for the underlying store is of string format:
+                //      <endpointId>_Pri<priority>
+                return endpointId + "_Pri" + priority.ToString();

```

Comment:
Do we know that endpointId defined in routes is case-insensitive. right?

Reply:
shouldn't matter either way, right?

Before my change, endpointId was just used directly, all I'm doing now is adding a fixed-format suffix.  So regardless of whether endpointId is case-sensitive or not, it should have no new change on the underlying storage.

---
In reply to: [391350197](https://github.com/Azure/iotedge/pull/2670#discussion_r391350197) [](ancestors = 391350197)